File: 15.三数之和.py
from cgitb import reset
from turtle import towards
from typing import *
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#

# @lc code=start
# https://mp.weixin.qq.com/s/fSyJVvggxHq28a0SdmZm6Q
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        length = len(nums)
        
        if length < 3:
            return []
        nums.sort()
        res = []
        for i in range(length):
            if i > 0 and nums[i-1] == nums[i]:
                continue

            two_ress = self.twoSum(nums, i+1, 0 - nums[i])
            for two_res in two_ress:
                two_res.append(nums[i])
                res.append(two_res)
        return res 
        
    def twoSum(self, nums: List[int], start: int, target: int) -> List[List[int]]:
        length = len(nums)
        if length < 2:
            return []
        low, high = start, length - 1
        ans = []

        while low < high:
            left_num = nums[low]
            right_num = nums[high]
            sum = left_num + right_num
            if sum < target:
                while low < high and nums[low] == left_num:
                    low += 1
            elif sum > target:
                while low < high and nums[high] == right_num:
                    high -= 1 
            else:
                ans.append([nums[low], nums[high]])
                while low < high and nums[low] == left_num:
                    low += 1
                while low < high and nums[high] == right_num:
                    high -= 1 
        return ans
# @lc code=end

a = Solution()
print(a.threeSum([-1,0,1,2,-1,-4]))

File: 15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start
// 此方法使用leecode官方方法
func threeSum(nums []int) [][]int {
	length := len(nums)
	sort.Ints(nums)
	res := make([][]int, 0)

	for first := 0; first < length; first++ {
		// 过滤和上个数一样的值，避免重复
		if first > 0 && nums[first] == nums[first-1] {
			continue
		}
		// 因为1 2都是从最小开始找，3就从最大开始找可以减小计算
		third := length - 1
		target := -1 * nums[first]
		for second := first + 1; second < length; second++ {
			// 过滤和上个数一样的值，避免重复
			if second > first+1 && nums[second] == nums[second-1] {
				continue
			}

			// 注意这里是while 因为third从大到校，基本上只会大于，小于等于说明没有直接跳过
			for second < third && nums[second]+nums[third] > target {
				third--
			}
			if second == third {
				break
			}
			if nums[second]+nums[third] == target {
				res = append(res, []int{nums[first], nums[second], nums[third]})
			}

		}
	}
	return res
}

// @lc code=end

